epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.kt

```diff
@@ -0,0 +1,498 @@
+package com.airbnb.epoxy
+
+import android.util.Log
+import android.util.SparseArray
+import android.view.View
+import androidx.annotation.IdRes
+import androidx.annotation.IntRange
+import androidx.recyclerview.widget.RecyclerView
+import com.airbnb.viewmodeladapter.R
+import java.util.ArrayList
+import java.util.HashMap
+
+/**
+ * A simple way to track visibility events on [com.airbnb.epoxy.EpoxyModel].
+ *
+ * [EpoxyVisibilityTracker] works with any [androidx.recyclerview.widget.RecyclerView]
+ * backed by an Epoxy controller. Once attached the events will be forwarded to the Epoxy model (or
+ * to the Epoxy view when using annotations).
+ *
+ * @see OnVisibilityChanged
+ *
+ * @see OnVisibilityStateChanged
+ *
+ * @see OnModelVisibilityChangedListener
+ *
+ * @see OnModelVisibilityStateChangedListener
+ */
+class EpoxyVisibilityTracker {
+
+    /**
+     * Used to listen to [RecyclerView.ItemAnimator] ending animations.
+     */
+    private val itemAnimatorFinishedListener =
+        RecyclerView.ItemAnimator.ItemAnimatorFinishedListener {
+            processChangeEvent(
+                "ItemAnimatorFinishedListener.onAnimationsFinished",
+                /* don't check item animator to prevent recursion */ false
+            )
+        }
+
+    /** Maintain visibility item indexed by view id (identity hashcode)  */
+    private val visibilityIdToItemMap = SparseArray<EpoxyVisibilityItem>()
+    private val visibilityIdToItems: MutableList<EpoxyVisibilityItem> = ArrayList()
+
+    /** listener used to process scroll, layout and attach events  */
+    private val listener = Listener()
+
+    /** listener used to process data events  */
+    private val observer = DataObserver()
+
+    private var attachedRecyclerView: RecyclerView? = null
+
+    private var lastAdapterSeen: RecyclerView.Adapter<*>? = null
+
+    private var onChangedEnabled = true
+
+    private var partialImpressionThresholdPercentage: Int? = null
+
+    /** All nested visibility trackers  */
+    private val nestedTrackers: MutableMap<RecyclerView, EpoxyVisibilityTracker> = HashMap()
+
+    /** This flag is for optimizing the process on detach. If detach is from data changed then it
+     * need to re-process all views, else no need (ex: scroll). */
+    private var visibleDataChanged = false
+
+    /**
+     * Enable or disable visibility changed event. Default is `true`, disable it if you don't need
+     * (triggered by every pixel scrolled).
+     *
+     * @see OnVisibilityChanged
+     *
+     * @see OnModelVisibilityChangedListener
+     */
+    fun setOnChangedEnabled(enabled: Boolean) {
+        onChangedEnabled = enabled
```

Comment:
is this for backwards compatibility? kind of a shame to have this function when we could just make the var public - wouldn't the java interop be the same with just the var?

Reply:
same for `setPartialImpressionThresholdPercentage` I guess

Reply:
Same ... it is IDE convert :)

Reply:
@elihart okay for a breaking change? `setPartialImpressionThresholdPercentage` won't be available from Kotlin, what is the general policy for this?

Reply:
@eboudrant I am in favor of a small breaking change like that (easy for people to fix) in order to set us up with a better API going forward :) thanks!